R/container.R

```diff
@@ -0,0 +1,174 @@
+#' Declare post-processing for model predictions
+#'
+#' @param mode The model's mode, one of `"unknown"`, `"classification"`, or
+#' `"regression"`. Modes of `"censored regression"` are not currently supported.
+#' @param type The model sub-type. Possible values are `"unknown"`, `"regression"`,
+#' `"binary"`, or `"multiclass"`.
+#' @param outcome The name of the outcome variable.
+#' @param estimate The name of the point estimate (e.g. predicted class)
+#' @param probabilities The names of class probability estimates (if any). For
+#' classification, these should be given in the order of the factor levels of
+#' the `estimate`.
+#' @param time The name of the predicted event time. (not yet supported)
+#' @param call The call to be displayed in warnings or errors.
+#' @examples
+#'
+#' container()
+#' @export
+container <- function(mode = "unknown", type = "unknown", outcome = character(0),
+                      estimate = character(0), probabilities = character(0),
+                      time = character(0), call = rlang::current_env()) {
+  dat <-
+    list(
+      outcome = outcome,
+      type = type,
+      estimate = estimate,
+      probabilities = probabilities,
+      time = time
+    )
+  new_container(
+    mode,
+    type,
+    operations = list(),
+    columns = dat,
+    ptype = tibble::tibble(),
+    call = call
+  )
+}
+
+new_container <- function(mode, type, operations, columns, ptype, call) {
+  mode <- rlang::arg_match0(mode, c("unknown", "regression", "classification", "censored regression"))
+
+  if ( mode == "regression" ) {
+    type <- "regression"
+  }
+
+  type <- rlang::arg_match0(type, c("unknown", "regression", "binary", "multiclass"))
+
+  if ( !is.list(operations) ) {
+    cli::cli_abort("The {.arg operations} argument should be a list.", call = call)
+  }
+
+  is_oper <- purrr::map_lgl(operations, ~ inherits(.x, "operation"))
+  if ( length(is_oper) > 0 & !any(is_oper) ) {
+    bad_oper <- names(is_oper)[!is_oper]
+    cli::cli_abort("The following {.arg operations} do not have the class \\
+                   {.val operation}: {bad_oper}.", call = call)
+  }
+
+  # validate operation order and check duplicates
+  validate_oper_order(operations, mode, call)
+
+
+  # check columns
+
+  res <- list(mode = mode, type = type, operations = operations,
+              columns = columns, ptype = ptype)
+  class(res) <- "container"
+  res
+}
+
+#' @export
+print.container <- function(x, ...) {
+  # todo emulate Emil's recipe printing
+
+  num_op <- length(x$operations)
+  cli::cli_inform("{x$type} post-processing object with {num_op} operation{?s}")
+
+  if (num_op > 0) {
+    cat("\n")
+    res <- purrr::map(x$operations, ~ print(.x))
+  }
+
+  invisible(x)
+}
+
+
+# ------------------------------------------------------------------------------
+
+#' @export
+fit.container <- function(object, .data, outcome, estimate, probabilities = c(),
```

Comment:
How does `.data` as an argument grab you?

Reply:
I'd vote we just use `data` for consistency with `fit.model_spec` and `fit.workflow`.

Reply:
`data` does kind of stink in that it surfaces `data()` when not supplied, but I think consistency here outweighs that oddity.

Reply:
Yeah. I'm trying to avoid error messages about unsettable closures.